### PR TITLE
fix(kb): store kb cookie  in memory and manage new version

### PR DESCRIPTION
**Pull request by garnier-quentin**
## Description

Fix https://github.com/centreon/centreon/pull/8124
Fix https://github.com/centreon/centreon/issues/7790

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x
- [x] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

Set wrong knowledge base information. You get cookies in /tmp each 5 minutes. 

## Checklist

- [X] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/class/centreon-knowledge/wikiApi.class.php [hunk 1]
```diff
@@ -292,10 +265,19 @@ public function getAllPages()
...
             // Get next page if exists
-            if (isset($result->{'query-continue'}->allpages->apcontinue)) {
-                $postfields['apfrom'] = $result->{'query-continue'}->allpages->apcontinue;
+            $continue = false;
+            if ($this->version >= 1.31) {
+                if (isset($result->{'continue'}->apcontinue)) {
+                    $postfields['apfrom'] = $result->{'continue'}->apcontinue;
+                    $continue = true;
+                }
+            } else {
```
Comment (sc979) [inline www/class/centreon-knowledge/wikiApi.class.php:274]: you can merge the else with the next if

File: www/class/centreon-knowledge/wikiApi.class.php [hunk 2]
```diff
@@ -271,11 +271,9 @@ public function getAllPages()
                     $postfields['apfrom'] = $result->{'continue'}->apcontinue;
                     $continue = true;
                 }
-            } else {
-                if (isset($result->{'query-continue'}->allpages->apcontinue)) {
-                    $postfields['apfrom'] = $result->{'query-continue'}->allpages->apcontinue;
-                    $continue = true;
-                }
+            } else if (isset($result->{'query-continue'}->allpages->apcontinue)) {
```
Comment (sc979) [inline www/class/centreon-knowledge/wikiApi.class.php:274]:
```suggestion
            } elseif (isset($result->{'query-continue'}->allpages->apcontinue)) {
```

File: www/class/centreon-knowledge/wikiApi.class.php [hunk 3]
```diff
@@ -71,12 +71,10 @@ public function __construct()
     private function getCurl()
     {
         $curl = curl_init();
-        $cookiefile = tempnam("/tmp", "CURLCOOKIE");
         curl_setopt($curl, CURLOPT_URL, $this->url);
         curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
         curl_setopt($curl, CURLOPT_POST, true);
-        curl_setopt($curl, CURLOPT_COOKIEFILE, $cookiefile);
-        curl_setopt($curl, CURLOPT_COOKIEJAR, $cookiefile);
+        curl_setopt($curl, CURLOPT_COOKIEFILE, '');
```
Comment (sc979) [inline www/class/centreon-knowledge/wikiApi.class.php:77]: why keep the CookieFile option, when it's empty ? shouldn't it be removed ?

Comment (garnier-quentin) [reply www/class/centreon-knowledge/wikiApi.class.php:77]: Empty means: cookies are in memory

File: www/class/centreon-knowledge/wikiApi.class.php [hunk 4]
```diff
@@ -150,27 +138,18 @@ public function login()
 
         curl_setopt($this->curl, CURLOPT_POSTFIELDS, $postfields);
         $result = curl_exec($this->curl);
+        if ($result == false) {
```
Comment (kduret) [inline www/class/centreon-knowledge/wikiApi.class.php:141]:
```suggestion
        if ($result === false) {
```

File: www/class/centreon-knowledge/wikiApi.class.php [hunk 5]
```diff
@@ -123,20 +119,12 @@ public function login()
...
-
-        // Get cookies
-        preg_match_all('/^Set-Cookie:\s*([^;]*)/mi', $header, $matches);
-
-        $this->cookies = array_merge($this->cookies, $matches[1]);
-        $cookies = implode('; ', $this->cookies);
-        curl_setopt($this->curl, CURLOPT_COOKIE, $cookies);
-
-        $result = json_decode($body, true);
+        if ($result == false) {
```
Comment (kduret) [inline www/class/centreon-knowledge/wikiApi.class.php:122]:
```suggestion
        if ($result === false) {
```

**Pull request merged**